src/ddspdrum/torchmodule.py

```diff
@@ -0,0 +1,406 @@
+                ModParameter("attack", a, 0.0, 20.0, curve="log"),
+                ModParameter("decay", d, 0.0, 20.0, curve="log"),
+                ModParameter("sustain", s, 0.0, 1.0),
+                ModParameter("release", r, 0.0, 20.0, curve="log"),
+                ModParameter("alpha", alpha, 0.0, 10.0),
+            ]
+        )
+
+    def forward(self, note_on_duration: T = T(0)) -> np.ndarray:
+        """Generate an ADSR envelope.
+
+        By default, this envelope reacts as if it was triggered with midi, for
+        example playing a keyboard. Each midi event has a beginning and end:
+        note-on, when you press the key down; and note-off, when you release the
+        key. `note_on_duration` is the amount of time that the key is depressed.
+
+        During the note-on, the envelope moves through the attack and decay
+        sections of the envelope. This leads to musically-intuitive, but
+        programatically-counterintuitive behaviour:
+
+        E.g., assume attack is .5 seconds, and decay is .5 seconds. If a note is
+        held for .75 seconds, the envelope won't pass through the entire
+        attack-and-decay (specifically, it will execute the entire attack, and
+        only .25 seconds of the decay).
+
+        Alternately, you can specify a `note_on_duration` of "0" which will
+        switch the envelope to one-shot mode. In this case, the envelope moves
+        through the entire attack, decay, and release, with no held "sustain"
+        value.
+
+        If this is confusing, don't worry about it. ADSR's do a lot of work
+        behind the scenes to make the playing experience feel natural.
+
+        """
+
+        assert note_on_duration >= 0
+
+        # If sustain is "0" go to one-shot mode (moves through ADR sections).
+        if note_on_duration == T(0):
+            note_on_duration = self.p("attack") + self.p("decay")
+
+        num_samples = self.seconds_to_samples(note_on_duration)
+
+        # Release decays from the last value of the attack-and-decay sections.
+        ADS = self.note_on(num_samples)
+        R = self.note_off(ADS[-1])
+
+        return torch.cat((ADS, R))
+
+    def _ramp(self, duration: T):
+        """Makes a ramp of a given duration in seconds.
+
+        This function is used for the piece-wise construction of the envelope
+        signal. Its output monotonically increases from 0 to 1. As a result,
+        each component of the envelope is a scaled and possibly reversed
+        version of this ramp:
+
+        attack      -->     returns an `a`-length ramp, as is.
+        decay       -->     `d`-length reverse ramp, descends from 1 to `s`.
+        release     -->     `r`-length reverse ramp, descends to 0.
+
+        Its curve is determined by alpha:
+
+        alpha = 1 --> linear,
+        alpha > 1 --> exponential,
+        alpha < 1 --> logarithmic.
+
+        """
+
+        # TODO: We originally used endpoint=False, which torch.linspace doesn't support :(
```

Comment:
@maxsolomonhenry thoughts?

Reply:
Yeah just looking into this. We could make a wrapper for torch linspace that does the extra calculations to ignore the endpoint. Or we could just plug our ears and hope this doesn't make a big difference.